text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+this could be very confusing to users since these nodes would likely have to be
+text nodes (how Glimmer tracks component boundaries internally), and would force
+users to write unintuitive looping logic to traverse nodes:
+
+```js
+let { firstNode, lastNode } = bounds;
+let currentNode = firstNode;
+
+while (currentNode !== lastNode) {
+  // do something with current node
+  currentNode = currentNode.nextSibling;
+}
+```
+
+This exposes users to low level APIs which may be useful to some, but in general
+are too complicated and nuanced. Instead, we can use new element modifiers which
+allow users to run code declaratively on the element:
+
+```hbs
+<div {{on-insert this.didInsertElement}} ...attributes>
+  First root element
+</div>
+<div {{on-insert this.didInsertElement}}>
+  Second root element
+</div>
+```
+
+This allows users to target which elements they care about directly. It also
+allows for code to be run on elements which are contained in conditionals,
+meaning users do not need to add messy logic to track component state
+themselves:
+
+```hbs
+{{#if shouldRender}}
+  <div
+    {{on-insert this.didInsertElement}}
+    {{on-destroy this.willDestroyElement}}
+  >
+
+  </div>
+{{/if}}
+```
+
+Finally, it means that users can bypass component code altogether by using
+helpers:
+
+```hbs
+<div {{on-insert (setScrollPosition @savedScrollPosition)}}></div>
+```
+
+This RFC proposes adding two element modifiers:
+
+1. `{{on-insert}}`, which runs immediately after the element is inserted.
+   `{{on-insert}}` is guaranteed to run before the enclosing component's
+   `didCreate` and `didUpdate` hooks, and after any child element's
+   `{{on-insert}}` modifiers, preserving the semantics of `didInsertElement`.
+
+   `{{on-insert}}` receives a function with the following signature:
+
+   ```ts
+   type OnInsertHandler = function(element: Element): void;
+   ```
+
+2. `{{on-destroy}}`, which runs immediately before the element is removed.
+   `{{on-destroy}}` is guaranteed to run before the enclosing component's
+   `didUpdate` and `destroy` hooks, and after and child element's
+   `{{on-destroy}}` modifiers, preserving the semantics of `willDestroyElement`.
+
+   `{{on-destroy}}` receives a function with the following signature:
+
+   ```ts
+   type OnDestroyHandler = function(element: Element): void;
+   ```
+
+#### Default Binding of Functions
+
+Currently, the `{{action}}` modifier is the only modifier which ships with
+Ember, and `{{action}}` by default binds functions to the context of the
+_component_. With nested values, this behavior does not particularly make sense:
+
+```hbs
+<div {{on-insert this.animationService.registerElement}}></div>
+```
+
+In this example, `on-insert` would effectively be running
+`this.animationService.registerElement.call(this, element)`, where `this` refers
+to the _component_ and not the _animation service_.
+
+The `bind` RFC, which runs parallel to this one, proposes that a new helper with
```

Comment:
if this RFC is already up, we should link to it.

Reply:
It is not yet written, but it will be up and published before this _if_ we end up going through with it. Going to be thinking it over in the next few days.